_posts/2017-12-19-fuzzy-search-in-elixir.md

```diff
@@ -0,0 +1,284 @@
+---
+layout: post
+title:  "Implementing Fuzzy Search in Elixir Using Ecto and Postgres"
+date:   2017-12-19 12:00:00 -0400
+categories: api
+author: Demi Filippou
+---
+
+## Background
+Here at Forward Financing, our Elixir single sign on app has always stored only a small number of users. We developed only internal tools and therefore only kept track of internal users. However, now that we are expanding and developing external facing tools, the number of users is skyrocketing. This caused a problem for admins who were creating, editing, and checking for users. To check if a user exists, an admin would have to manually click through the growing number of pages of users. Yuck. That is why this sprint, our project managers decided it was time to implement search functionality for users. We decided to go down the route of fuzzy search, because why vanilla search when you can fuzzy search?
+
+### Where do we start?
+If we are going to write fuzzy search, we are going to need a way to calculate what strings are close enough to our query to show up in the results. After some quick investigating, we decided the best route to go down was to use the Levenshtein function from the [fuzzystrmatch Postgres module](https://www.postgresql.org/docs/current/static/fuzzystrmatch.html#idm46428636025488)
+
+### Levenshtein? What is that? Why are we using it?
+Levenshtein is often referred to as "edit distance," a name that better reflects its purpose. The Levenshtein distance between two strings is essentially the number of character changes to transform string A to string B. Valid character changes are:
+* Adding a character
+* Deleting a character
+* Changing a character
+
+So a low Levenshtein distance means two strings are very similar. This makes it the perfect tool for fuzzy search because we can use it to find results similar to the search query. This will make our search tolerant to typos and small misspellings.
+
+## The Solution
+### Migration
+The first thing we want to do is add our fuzzystrmatch extension. We'll need a migration for this.
+
+`mix ecto.gen.migration add_fuzzystrmatch_extension`
+
+Make sure to define `up` and `down` rather than defining the default `change` Elixir provides, or Ecto will be unable to rollback this migration.
+```elixir
+defmodule MyApp.Migrations.AddFuzzystrmatchExtension do
+  use Ecto.Migration
+
+  def up do
+    execute "CREATE extension if not exists fuzzystrmatch;"
+  end
+
+  def down do
+    execute "DROP extension if exists fuzzystrmatch;"
+  end
+end
+```
+Now we're ready for the fun part.
+
+### Using Levenshtein
+I started off simple. Just a SQL query that we will execute using [Ecto's SQL adapter](https://hexdocs.pm/ecto/Ecto.Adapters.SQL.html#query/4). We will only search on email for now for simplicity. Let's add it to our model.
```

Comment:
simple~.~ `--` Just

Reply:
`For now` We will only search on email ~for now~ for simplicity. 

Too many "fors" close together